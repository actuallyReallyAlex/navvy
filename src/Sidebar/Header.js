import React from 'react'
import PropTypes from 'prop-types'
import { Text } from 'grommet'

const Header = ({ heading, size }) => {
  return (
    <Text
      margin={{ horizontal: 'small', top: 'medium' }}
      size="small"
      style={{ textAlign: size === 'expanded' ? 'start' : 'center' }}
    >
      {heading}
    </Text>
  )
}

Header.propTypes = {
  heading: PropTypes.string.isRequired,
  size: PropTypes.oneOf(['expanded', 'collapsed']).isRequired
}

export default Header
