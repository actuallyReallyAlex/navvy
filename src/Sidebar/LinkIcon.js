import React from 'react'
import PropTypes from 'prop-types'
import * as grommetIcons from 'grommet-icons'

const LinkIcon = ({ icon }) => {
  const Icon = grommetIcons[icon]
  return <Icon />
}

LinkIcon.propTypes = {
  icon: PropTypes.string.isRequired
}

export default LinkIcon
