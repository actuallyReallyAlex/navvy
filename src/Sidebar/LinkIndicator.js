import React from 'react'
import { Box } from 'grommet'

const LinkIndicator = () => {
  return (
    <Box
      background="blue"
      round="large"
      width="5px"
      style={{
        position: 'absolute',
        left: '0',
        right: '5px',
        top: '5px',
        bottom: '5px'
      }}
    />
  )
}

export default LinkIndicator
