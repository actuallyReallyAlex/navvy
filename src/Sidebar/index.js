import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Box } from 'grommet'
import Logo from './Logo'
import Section from './Section'
import { Scrollbars } from 'react-custom-scrollbars'

const Sidebar = ({ background, sections, size }) => {
  const firstLinkTitle = sections[0].links[0].title
  const [selected, setSelected] = useState(firstLinkTitle)

  const sizes = {
    expanded: '300px',
    collapsed: '100px'
  }

  return (
    <Box
      background={`${background}-3`}
      fill="vertical"
      style={{ transition: 'all 0.5s' }}
      width={sizes[size]}
    >
      <Logo size={size} />
      <Scrollbars autoHide>
        {sections.map(({ heading, links }) => (
          <Section
            heading={heading}
            key={heading}
            links={links}
            selected={selected}
            setSelected={setSelected}
            size={size}
          />
        ))}
      </Scrollbars>
    </Box>
  )
}

Sidebar.propTypes = {
  background: PropTypes.oneOf(['dark', 'light']),
  sections: PropTypes.array.isRequired,
  size: PropTypes.oneOf(['expanded', 'collapsed']).isRequired
}

export default Sidebar
