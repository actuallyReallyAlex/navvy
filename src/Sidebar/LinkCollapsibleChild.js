import React from 'react'
import PropTypes from 'prop-types'
import { Box, Button, Text } from 'grommet'
import LinkIcon from './LinkIcon'

const LinkCollapsibleChild = ({ handler, icon, size, title }) => {
  const handleClick = () => {
    handler && handler()
  }

  return (
    <Button hoverIndicator onClick={handleClick} plain>
      <Box
        background="light-5"
        direction="row"
        height="45px"
        margin={{ vertical: 'none', left: 'medium' }}
        pad={{ vertical: 'none', left: 'small' }}
        style={{ position: 'relative' }}
        width="88%"
      >
        <Box
          align="center"
          direction="row"
          gap="small"
          justify="start"
          width="100%"
        >
          <LinkIcon icon={icon} />
          <Text>{title}</Text>
        </Box>
      </Box>
    </Button>
  )
}

LinkCollapsibleChild.propTypes = {
  handler: PropTypes.func,
  icon: PropTypes.string,
  size: PropTypes.oneOf(['expanded', 'collapsed']).isRequired,
  title: PropTypes.string.isRequired
}

export default LinkCollapsibleChild
