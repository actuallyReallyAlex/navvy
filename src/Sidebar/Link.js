import React from 'react'
import PropTypes from 'prop-types'
import { Box, Button, Text } from 'grommet'
import LinkIndicator from './LinkIndicator'
import LinkIcon from './LinkIcon'

const Link = ({ handler, icon, selected, setSelected, size, title }) => {
  const handleClick = () => {
    setSelected(title)
    handler && handler()
  }

  return (
    <Button hoverIndicator onClick={handleClick} plain>
      <Box
        direction="row"
        height="45px"
        margin={{ vertical: 'xsmall', left: 'xsmall' }}
        pad={{ vertical: 'small', left: 'small' }}
        style={{ position: 'relative' }}
        width="95%"
      >
        {selected === title && <LinkIndicator />}
        <Box
          align="center"
          direction="row"
          gap="small"
          justify={size === 'expanded' ? 'start' : 'center'}
          width={size === 'expanded' ? '100%' : '75%'}
        >
          <LinkIcon icon={icon} />
          {size === 'expanded' && <Text>{title}</Text>}
        </Box>
      </Box>
    </Button>
  )
}

Link.propTypes = {
  handler: PropTypes.func,
  icon: PropTypes.string,
  selected: PropTypes.string.isRequired,
  setSelected: PropTypes.func.isRequired,
  size: PropTypes.oneOf(['expanded', 'collapsed']).isRequired,
  title: PropTypes.string.isRequired
}

Link.defaultProps = {
  icon: 'FormSubtract'
}

export default Link
