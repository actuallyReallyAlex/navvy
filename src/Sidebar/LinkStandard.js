import React from 'react'
import PropTypes from 'prop-types'
import { Box, Button, Text } from 'grommet'
import LinkIndicator from './LinkIndicator'
import LinkIcon from './LinkIcon'

const LinkStandard = ({
  handler,
  icon,
  selected,
  setSelected,
  size,
  title
}) => {
  const handleClick = () => {
    setSelected(title)
    handler && handler()
  }

  return (
    <Button hoverIndicator onClick={handleClick} plain>
      <Box
        direction="row"
        height="45px"
        margin={{ vertical: 'xsmall', left: 'xsmall' }}
        pad={{ vertical: 'small', left: 'small' }}
        style={{ position: 'relative' }}
        width="95%"
      >
        {selected === title && <LinkIndicator />}
        <Box
          align="center"
          direction="row"
          gap="small"
          justify={size === 'expanded' ? 'start' : 'center'}
          width={size === 'expanded' ? '100%' : '75%'}
        >
          <LinkIcon icon={icon} />
          {size === 'expanded' && <Text>{title}</Text>}
        </Box>
      </Box>
    </Button>
  )
}

LinkStandard.propTypes = {
  handler: PropTypes.func,
  icon: PropTypes.string,
  selected: PropTypes.string.isRequired,
  setSelected: PropTypes.func.isRequired,
  size: PropTypes.oneOf(['expanded', 'collapsed']).isRequired,
  title: PropTypes.string.isRequired
}

export default LinkStandard
