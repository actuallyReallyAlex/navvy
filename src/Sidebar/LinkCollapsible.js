import React, { Fragment, useState } from 'react'
import PropTypes from 'prop-types'
import { Box, Button, Text } from 'grommet'
import { FormNext, FormDown } from 'grommet-icons'
import LinkIndicator from './LinkIndicator'
import LinkIcon from './LinkIcon'
import LinkCollapsibleChild from './LinkCollapsibleChild'

const LinkCollapsible = ({
  children,
  handler,
  icon,
  selected,
  setSelected,
  size,
  title
}) => {
  const [expanded, setExpanded] = useState(false)

  const handleClick = () => {
    setSelected(title)
    setExpanded(!expanded)
    handler && handler()
  }

  return (
    <Fragment>
      <Button hoverIndicator onClick={handleClick} plain>
        <Box
          background={expanded ? 'light-5' : undefined}
          direction="row"
          height="45px"
          margin={
            expanded
              ? { top: 'xsmall', bottom: 'none', left: 'xsmall' }
              : { vertical: 'xsmall', left: 'xsmall' }
          }
          pad={{ vertical: 'small', left: 'small' }}
          style={{ position: 'relative' }}
          width="95%"
        >
          {selected === title && <LinkIndicator />}
          <Box
            align="center"
            direction="row"
            gap="small"
            justify={size === 'expanded' ? 'start' : 'center'}
            width={size === 'expanded' ? '100%' : '75%'}
          >
            <LinkIcon icon={icon} />
            {size === 'expanded' && <Text>{title}</Text>}
          </Box>
          <Box align="center" justify="center" pad={{ horizontal: 'small' }}>
            {expanded ? <FormDown /> : <FormNext />}
          </Box>
        </Box>
      </Button>
      {expanded &&
        children.map(({ handler, icon, title }, i) => (
          <LinkCollapsibleChild
            handler={handler}
            key={i}
            icon={icon}
            size={size}
            title={title}
          />
        ))}
    </Fragment>
  )
}

LinkCollapsible.propTypes = {
  children: PropTypes.array.isRequired,
  handler: PropTypes.func,
  icon: PropTypes.string,
  selected: PropTypes.string.isRequired,
  setSelected: PropTypes.func.isRequired,
  size: PropTypes.oneOf(['expanded', 'collapsed']).isRequired,
  title: PropTypes.string.isRequired
}

export default LinkCollapsible
