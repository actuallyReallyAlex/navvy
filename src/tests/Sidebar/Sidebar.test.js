import React from 'react'
import Sidebar from '../../Sidebar/index'
import { shallow } from 'enzyme'

const defaultProps = {
  sections: [
    {
      heading: 'Navigation',
      links: [
        { title: 'Dashboard' },
        {
          title: 'Page layouts',
          handler: () => console.log('LOL')
        },
        { title: 'Navigation' },
        { title: 'Widget' }
      ]
    },
    {
      heading: 'UI Element',
      links: [
        { title: 'Basic' },
        { title: 'Advance' },
        { title: 'Extra' },
        { title: 'Animations' },
        { title: 'Icons' }
      ]
    },
    {
      heading: 'Forms',
      links: [
        { title: 'Form' },
        { title: 'Form Picker' },
        { title: 'Form Select' },
        { title: 'Form Masking' },
        { title: 'Form Wizard' }
      ]
    }
  ]
}

describe('<Sidebar />', () => {
  afterEach(() => {
    // defaultProps.handler.mockClear()
  })

  test('Should render the <Sidebar /> component.', () => {
    const component = shallow(<Sidebar {...defaultProps} />)
    expect(component).toMatchSnapshot()
  })

  test('Should be able to render in "light" mode.', () => {
    const component = shallow(<Sidebar {...defaultProps} background="light" />)
    expect(component).toMatchSnapshot()
  })

  test('Should be able to render in "dark" mode.', () => {
    const component = shallow(<Sidebar {...defaultProps} background="dark" />)
    expect(component).toMatchSnapshot()
  })

  test('Should have the ability to collapse the sidebar.', () => {
    const component = shallow(<Sidebar {...defaultProps} />)
    const collapseButton = component.find('collapse-button-selector')
    expect(component).toMatchSnapshot()
    collapseButton.simulate('click')
    expect(component).toMatchSnapshot()
  })
})
