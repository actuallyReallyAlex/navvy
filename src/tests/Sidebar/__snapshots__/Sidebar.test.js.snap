// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> Should be able to render in "dark" mode. 1`] = `ShallowWrapper {}`;

exports[`<Sidebar /> Should be able to render in "light" mode. 1`] = `ShallowWrapper {}`;

exports[`<Sidebar /> Should render the <Sidebar /> component. 1`] = `ShallowWrapper {}`;
